File: src/PythonClient/lib/quixstreaming/streamreader.py
from .eventhook import EventHook
from .models import *
from . import __importnet
import Quix.Sdk.Streaming

from .models.streampackage import StreamPackage
from .models.streamreader import StreamPropertiesReader, StreamParametersReader
from .models.streamreader.streameventsreader import StreamEventsReader


class StreamReader(object):
    """
        Handles reading stream from a topic
    """

    def __init__(self, net_object: Quix.Sdk.Streaming.IStreamReader):
        """
        Initializes a new instance of StreamReader.
        NOTE: Do not initialize this class manually, use StreamingClient.on_stream_received to read streams
        :param net_object: (.net object): The .net object representing a StreamReader
        """
        self.__wrapped = net_object  # the wrapped .net StreamReader

        self.on_stream_closed = EventHook(name="StreamReader.on_stream_closed")
        """Raised when stream close is read.       
         Parameters:
            sender (StreamReader): The StreamReader to which the event belongs to.            
            end_type (StreamEndType): The StreamEndType value read from the message.
        """

        def __on_stream_closed_net_handler(sender, end_type):
            self.on_stream_closed.fire(StreamEndType.convert_from_net(end_type))
        self.__wrapped.OnStreamClosed += __on_stream_closed_net_handler

        def __on_package_received_net_handler(sender, package):
            self.on_package_received.fire(self, StreamPackage(package))

        def __on_first_sub_on_package_received():
            self.__wrapped.OnPackageReceived += __on_package_received_net_handler

        def __on_last_unsub_on_package_received():
            self.__wrapped.OnPackageReceived -= __on_package_received_net_handler

        self.on_package_received = EventHook(__on_first_sub_on_package_received, __on_last_unsub_on_package_received, name="StreamReader.on_package_received")
        """Raised when stream package has been received.       
         Parameters:
            sender (StreamReader): The StreamReader to which the event belongs to.            
            package (StreamPackage): The Package received.
        """


    @property
    def stream_id(self) -> str:
        """Get the id of the stream being read"""
        return self.__wrapped.StreamId

    def dispose(self):
        """
            Disposes underlying .net instance.
        """
        self.__wrapped.Dispose()

    @property
    def properties(self) -> StreamPropertiesReader:
        """ Gets the reader for accessing the properties and metadata of the stream """
        return StreamPropertiesReader(self.__wrapped.Properties)

    @property
    def events(self) -> StreamEventsReader:
        """
        Gets the reader for accessing event related information of the stream such as definitions and event values
        """
        return StreamEventsReader(self.__wrapped.Events)

    @property
    def parameters(self) -> StreamParametersReader:
        """
        Gets the reader for accessing parameter related information of the stream such as definitions and parameter values
        """
        return StreamParametersReader(self.__wrapped.Parameters)
        

File: src/CsharpClient/Quix.Sdk.Streaming/StreamReader.cs
using System;
using Microsoft.Extensions.Logging;
using Quix.Sdk.Process;
using Quix.Sdk.Process.Models;
using Quix.Sdk.Streaming.Models.StreamReader;

namespace Quix.Sdk.Streaming
{
    /// <summary>
    /// Handles reading data for the assigned stream from the protocol.
    /// </summary>
    internal class StreamReader : StreamProcess, IStreamReaderInternal
    {
        private readonly IInputTopic topic;
        private readonly ILogger logger = Logging.CreateLogger<StreamReader>();
        private readonly StreamPropertiesReader streamPropertiesReader;
        private readonly StreamParametersReader streamParametersReader;
        private readonly StreamEventsReader streamEventsReader;
        private bool isClosed = false;

        /// <summary>
        /// Initializes a new instance of <see cref="StreamReader"/>
        /// This constructor is called internally by the Stream Process factory.
        /// </summary>
        /// <param name="topic">The topic the reader belongs to</param>
        /// <param name="streamId">Stream Id of the source that has generated this Stream Process. 
        /// Commonly the Stream Id will be coming from the protocol. 
        /// If no stream Id is passed, like when a new stream is created for producing data, a Guid is generated automatically.</param>
        internal StreamReader(IInputTopic topic, string streamId): base(streamId)
        {
            this.topic = topic;
            // Managed readers
            this.streamPropertiesReader = new StreamPropertiesReader(this.topic, this);
            this.streamParametersReader = new StreamParametersReader(this.topic, this);
            this.streamEventsReader = new StreamEventsReader(this.topic, this);

            InitializeStreaming();
        }

        /// <summary>
        /// Exists for mocking purposes
        /// </summary>
        protected StreamReader()
        {
            
        }

        /// <inheritdoc />
        public StreamPropertiesReader Properties => streamPropertiesReader;

        /// <inheritdoc />
        public StreamParametersReader Parameters => streamParametersReader;

        /// <inheritdoc />
        public StreamEventsReader Events => streamEventsReader;

        /// <inheritdoc />
        public event EventHandler<PackageReceivedEventArgs> OnPackageReceived;

        /// <inheritdoc />
        public event EventHandler<StreamClosedEventArgs> OnStreamClosed;


        /// <inheritdoc />
        public virtual event Action<IStreamReader, Process.Models.StreamProperties> OnStreamPropertiesChanged;

        /// <inheritdoc />
        public virtual event Action<IStreamReader, Process.Models.ParameterDefinitions> OnParameterDefinitionsChanged;

        /// <inheritdoc />
        public virtual event Action<IStreamReader, Process.Models.TimeseriesDataRaw> OnTimeseriesData;

        /// <inheritdoc />
        public virtual  event Action<IStreamReader, Process.Models.EventDataRaw> OnEventData;

        /// <inheritdoc />
        public virtual  event Action<IStreamReader, Process.Models.EventDefinitions> OnEventDefinitionsChanged;

        private void InitializeStreaming()
        {
            // Modifiers
            // this.AddComponent(SimpleModifier)

            this.Subscribe<Process.Models.StreamProperties>(OnStreamPropertiesReceived);
            this.Subscribe<Process.Models.TimeseriesDataRaw>(OnTimeseriesDataReceived);
            this.Subscribe<Process.Models.ParameterDefinitions>(OnParameterDefinitionsReceived);
            this.Subscribe<Process.Models.EventDataRaw[]>(OnEventDataReceived);
            this.Subscribe<Process.Models.EventDefinitions>(OnEventDefinitionsReceived);
            this.Subscribe<Process.Models.StreamEnd>(OnStreamEndReceived);
            this.Subscribe(OnStreamPackageReceived);

            this.OnClosed += () =>
            {
                RaiseStreamClosed(StreamEndType.Terminated);
            };
        }

        private void OnStreamPackageReceived(IStreamProcess streamProcess, Sdk.Process.Models.StreamPackage package)
        {
            this.logger.LogTrace("StreamReader: OnPackageReceived");
            this.OnPackageReceived?.Invoke(this, new PackageReceivedEventArgs(this.topic, this, package));
        }

        private void OnStreamPropertiesReceived(IStreamProcess streamProcess, Process.Models.StreamProperties obj)
        {
            this.logger.LogTrace("StreamReader: OnStreamPropertiesReceived");
            this.OnStreamPropertiesChanged?.Invoke(this, obj);
        }

        private void OnTimeseriesDataReceived(IStreamProcess streamProcess, Process.Models.TimeseriesDataRaw obj)
        {
            this.logger.LogTrace("StreamReader: OnTimeseriesDataReceived. Data packet of size = {0}", obj.Timestamps.Length);
            this.OnTimeseriesData?.Invoke(this, obj);
        }

        private void OnParameterDefinitionsReceived(IStreamProcess streamProcess, Process.Models.ParameterDefinitions obj)
        {
            this.logger.LogTrace("StreamReader: OnParameterDefinitionsReceived");
            this.OnParameterDefinitionsChanged?.Invoke(this, obj);
        }

        private void OnEventDataReceived(IStreamProcess streamProcess, Process.Models.EventDataRaw[] events)
        {
            this.logger.LogTrace("StreamReader: OnEventDataReceived");
            for (var index = 0; index < events.Length; index++)
            {
                var ev = events[index];
                this.OnEventData?.Invoke(this, ev);
            }
        }

        private void OnEventDefinitionsReceived(IStreamProcess streamProcess, Process.Models.EventDefinitions obj)
        {
            this.logger.LogTrace("StreamReader: OnEventDefinitionsReceived");
            this.OnEventDefinitionsChanged?.Invoke(this, obj);
        }

        private void OnStreamEndReceived(IStreamProcess streamProcess, Process.Models.StreamEnd obj)
        {
            RaiseStreamClosed(obj.StreamEndType);
        }

        private void RaiseStreamClosed(StreamEndType endType)
        {
            if (isClosed) return;
            isClosed = true;
            this.logger.LogTrace("StreamReader: OnStreamEndReceived");

            this.Parameters.Buffers.ForEach(buffer => buffer.Dispose());
            
            this.OnStreamClosed?.Invoke(this, new StreamClosedEventArgs(this.topic, this, endType));
        }

        public override void Dispose()
        {
            this.streamEventsReader.Dispose();
            this.streamParametersReader.Dispose();
            this.streamPropertiesReader.Dispose();
            base.Dispose();
        }
    }
}
